File: algorithms/3_hard/P/02435_paths_in_matrix_whose_sum_is_divisible_by_k/dp_iterative.cpp
const int MOD_TO{1'000'000'007};

class Solution {
public:
    static int numberOfPaths(vector<vector<int>>& grid, int k) {
        const size_t m{grid.size()};
        const size_t n{grid[0].size()};
        const size_t k_usize{static_cast<size_t>(k)};
        vector<vector<vector<int>>> dp(m, vector<vector<int>>(n, vector<int>(k_usize, 0)));

        dp[0][0][static_cast<size_t>(grid[0][0] % k)] = 1;

        for (size_t i{1}; i < m; ++i) {
            int val{grid[i][0]};

            for (int t{0}; t < k; ++t) {
                dp[i][0][static_cast<size_t>((t + val) % k)] = dp[i - 1][0][static_cast<size_t>(t)];
            }
        }

        for (size_t j{1}; j < n; ++j) {
            int val{grid[0][j]};

            for (int t{0}; t < k; ++t) {
                dp[0][j][static_cast<size_t>((t + val) % k)] = dp[0][j - 1][static_cast<size_t>(t)];
            }
        }

        for (size_t i{1}; i < m; ++i) {
            for (size_t j{1}; j < n; ++j) {
                int val{grid[i][j]};

                for (int t{0}; t < k; ++t) {
                    dp[i][j][static_cast<size_t>((t + val) % k)] = (dp[i][j - 1][static_cast<size_t>(t)] + 
                                                                    dp[i - 1][j][static_cast<size_t>(t)]) % MOD_TO;
                }
            }
        }

        return dp[m - 1][n - 1][0];
    }
};

File: algorithms/3_hard/P/02435_paths_in_matrix_whose_sum_is_divisible_by_k/dp_iterative.rs
impl Solution {
    const MOD_TO: i32 = 1_000_000_007;

    pub fn number_of_paths(grid: Vec<Vec<i32>>, k: i32) -> i32 {
        let m = grid.len();
        let n = grid[0].len();
        let mut dp = vec![vec![vec![0_i32; k as usize]; n]; m];

        dp[0][0][(grid[0][0] % k) as usize] = 1;

        for i in 1..m {
            let val = grid[i][0];

            for t in 0..k {
                dp[i][0][((t + val) % k) as usize] = dp[i - 1][0][t as usize];
            }
        }

        for j in 1..n {
            let val = grid[0][j];

            for t in 0..k {
                dp[0][j][((t + val) % k) as usize] = dp[0][j - 1][t as usize];
            }
        }

        for i in 1..m {
            for j in 1.. n {
                let val = grid[i][j];

                for t in 0..k {
                    dp[i][j][((t + val) % k) as usize] = (dp[i][j - 1][t as usize] + dp[i - 1][j][t as usize]) % Self::MOD_TO;
                }
            }
        }

        dp[m - 1][n - 1][0]
    }
}
